### PR TITLE
[SPARK-43851][SQL] Support LCA in grouping expressions

### PR DESCRIPTION
**Pull request by Hisoka-X**
### What changes were proposed in this pull request?
This is a PR to support LCA in grouping expressions.

### Why are the changes needed?
Add new feature for LCA

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
exist test

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @cloud-fan @MaxGekk @wangyum @anchovYu

**Comment by cloud-fan** [comment]
My idea is we should define a dependency here. When resolving group by alias, we should not resolve the grouping col if its corresponding agg expression contains `LateralColumnAliasReference`. We wait for the LCA rule to rewrite Aggregate first. Is it doable?

**Comment by cloud-fan** [comment]
It seems the problem is not well understood. Looking at the classdoc of the LCA rule
```
 * ** Example for Aggregate:
 * Before rewrite:
 * Aggregate [dept#14] [dept#14 AS a#12, 'a + 1, avg(salary#16) AS b#13, 'b + avg(bonus#17)]
 * +- Child [dept#14,name#15,salary#16,bonus#17]
 *
 * After phase 1:
 * Aggregate [dept#14] [dept#14 AS a#12, lca(a) + 1, avg(salary#16) AS b#13, lca(b) + avg(bonus#17)]
 * +- Child [dept#14,name#15,salary#16,bonus#17]
 *
 * After phase 2:
 * Project [dept#14 AS a#12, lca(a) + 1, avg(salary)#26 AS b#13, lca(b) + avg(bonus)#27]
 * +- Aggregate [dept#14] [avg(salary#16) AS avg(salary)#26, avg(bonus#17) AS avg(bonus)#27,dept#14]
 *    +- Child [dept#14,name#15,salary#16,bonus#17]
 *
 * Now the problem falls back to the lateral alias resolution in Project.
```

The key is that we leave LCA in the newly added Project above Aggregate. I think LCA in grouping expressions needs a very different solution. We probably need to add Project under Aggregate, to evaluate LCA before executing the grouping expressions.

Can we revert https://github.com/apache/spark/pull/41804 first and think of the full story later?

File: sql/core/src/test/resources/sql-tests/analyzer-results/column-resolution-aggregate.sql.out [hunk 1]
```diff
@@ -94,27 +94,23 @@ org.apache.spark.sql.AnalysisException
...
 -- !query analysis
-Project [lca#x, (lca#x + 1) AS col#x]
-+- Project [k#x, k#x AS lca#x]
-   +- Aggregate [k#x, (k#x + 1)], [k#x]
-      +- SubqueryAlias v1
-         +- View (`v1`, [a#x,b#x,k#x])
-            +- Project [cast(a#x as int) AS a#x, cast(b#x as int) AS b#x, cast(k#x as int) AS k#x]
-               +- SubqueryAlias t
-                  +- LocalRelation [a#x, b#x, k#x]
+Aggregate [k#x, (k#x + 1)], [k#x AS lca#x, (k#x + 1) AS col#x]
```
[the remark this answers is not in the document]
Comment (anchovYu) [reply sql/core/src/test/resources/sql-tests/analyzer-results/column-resolution-aggregate.sql.out:97]: If `k` is a non-deterministic expression, will it be evaluate twice separately in `lca` and `lca+1`?

[the remark this answers is not in the document]
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/analyzer-results/column-resolution-aggregate.sql.out:97]: The key of LCA is we should evaluate the LCA expression only once, that's why I said we need to add a Project below Aggregate, to evaluate the LCA expression before doing grouping.

I don't think it's an easy job, considering all kinds of cases: nested LCA, LCA with agg func, LCA with grouping expr, etc. I suggest to revert the commit if we can't give a clear solution.

**Pull request closed without merging**